diff --git a/CRM/Contact/BAO/Contact/Permission.php b/CRM/Contact/BAO/Contact/Permission.php
index 731958c98a5..d5e16639287 100644
--- a/CRM/Contact/BAO/Contact/Permission.php
+++ b/CRM/Contact/BAO/Contact/Permission.php
@@ -236,20 +236,47 @@ public static function cache($userID, $type = CRM_Core_Permission::VIEW, $force
       }
     }
 
+    // grab a lock so other processes don't compete and do the same query
+    $lock = Civi::lockManager()->acquire("data.core.aclcontact.{$userID}");
+    if (!$lock->isAcquired()) {
+      // this can cause inconsistent results since we don't know if the other process
+      // will fill up the cache before our calling routine needs it.
+      // The default 3 second timeout should be enough for the other process to finish.
+      // However this routine does not return the status either, so basically
+      // its a "lets return and hope for the best"
+      return;
+    }
+
     $tables = [];
     $whereTables = [];
 
     $permission = CRM_ACL_API::whereClause($type, $tables, $whereTables, $userID, FALSE, FALSE, TRUE);
 
     $from = CRM_Contact_BAO_Query::fromClause($whereTables);
-    CRM_Core_DAO::executeQuery("
-INSERT INTO civicrm_acl_contact_cache ( user_id, contact_id, operation )
-SELECT DISTINCT $userID as user_id, contact_a.id as contact_id, '{$operation}' as operation
-         $from
-         LEFT JOIN civicrm_acl_contact_cache ac ON ac.user_id = $userID AND ac.contact_id = contact_a.id AND ac.operation = '{$operation}'
+
+    /* Ends up something like this:
+    CREATE TEMPORARY TABLE civicrm_temp_acl_contact_cache1310 (SELECT DISTINCT 2960 as user_id, contact_a.id as contact_id, 'View' as operation
+    FROM civicrm_contact contact_a  LEFT JOIN civicrm_group_contact_cache `civicrm_group_contact_cache-ACL` ON contact_a.id = `civicrm_group_contact_cache-ACL`.contact_id
+    LEFT JOIN civicrm_acl_contact_cache ac ON ac.user_id = 2960 AND ac.contact_id = contact_a.id AND ac.operation = 'View'
+    WHERE     ( `civicrm_group_contact_cache-ACL`.group_id IN (14, 25, 46, 47, 48, 49, 50, 51) )  AND (contact_a.is_deleted = 0)
+    AND ac.user_id IS NULL
+    */
+    /*$sql = "SELECT DISTINCT $userID as user_id, contact_a.id as contact_id, '{$operation}' as operation
+    $from
+    LEFT JOIN civicrm_acl_contact_cache ac ON ac.user_id = $userID AND ac.contact_id = contact_a.id AND ac.operation = '{$operation}'
+    WHERE    $permission
+    AND ac.user_id IS NULL
+    ";*/
+
+    $sql = "SELECT DISTINCT $userID as user_id, contact_a.id as contact_id, '{$operation}' as operation
+     $from
 WHERE    $permission
-AND ac.user_id IS NULL
-");
+";
+    $tempTable = 'civicrm_temp_acl_contact_cache' . rand(0, 2000);
+    $insertSql = "CREATE TEMPORARY TABLE $tempTable ($sql);";
+    CRM_Core_DAO::executeQuery($insertSql);
+    CRM_Core_DAO::executeQuery("INSERT IGNORE INTO civicrm_acl_contact_cache (user_id, contact_id, operation) SELECT user_id, contact_id, operation FROM $tempTable");
+    CRM_Core_DAO::executeQuery(" DROP TEMPORARY TABLE $tempTable");
 
     // Add in a row for the logged in contact. Do not try to combine with the above query or an ugly OR will appear in
     // the permission clause.
@@ -257,10 +284,11 @@ public static function cache($userID, $type = CRM_Core_Permission::VIEW, $force
       ($type == CRM_Core_Permission::VIEW && CRM_Core_Permission::check('view my contact'))) {
       if (!CRM_Core_DAO::singleValueQuery("
         SELECT count(*) FROM civicrm_acl_contact_cache WHERE user_id = %1 AND contact_id = %1 AND operation = '{$operation}' LIMIT 1", $queryParams)) {
-        CRM_Core_DAO::executeQuery("INSERT INTO civicrm_acl_contact_cache ( user_id, contact_id, operation ) VALUES(%1, %1, '{$operation}')", $queryParams);
+        CRM_Core_DAO::executeQuery("INSERT IGNORE INTO civicrm_acl_contact_cache ( user_id, contact_id, operation ) VALUES(%1, %1, '{$operation}')", $queryParams);
       }
     }
     Civi::$statics[__CLASS__]['processed'][$type][$userID] = 1;
+    $lock->release();
   }
 
   /**
diff --git a/CRM/Contact/BAO/GroupContact.php b/CRM/Contact/BAO/GroupContact.php
index c4540d79d53..6a2f86ad6eb 100644
--- a/CRM/Contact/BAO/GroupContact.php
+++ b/CRM/Contact/BAO/GroupContact.php
@@ -148,7 +148,16 @@ public static function addContactsToGroup(
     list($numContactsAdded, $numContactsNotAdded)
       = self::bulkAddContactsToGroup($contactIds, $groupId, $method, $status, $tracking);
 
-    CRM_Contact_BAO_Contact_Utils::clearContactCaches();
+    // Update group contact cache
+    $group = new CRM_Contact_BAO_Group();
+    $group->id = $groupId;
+    $group->find(TRUE);
+    CRM_Contact_BAO_GroupContactCache::load($group, TRUE);
+
+    // Update contact ACL cache
+    foreach ($contactIds as $contactId) {
+      CRM_ACL_BAO_Cache::updateEntry($contactId);
+    }
 
     CRM_Utils_Hook::post('create', 'GroupContact', $groupId, $contactIds);
 
@@ -242,7 +251,16 @@ public static function removeContactsFromGroup(
       }
     }
 
-    CRM_Contact_BAO_Contact_Utils::clearContactCaches();
+    // Update group contact cache
+    $group = new CRM_Contact_BAO_Group();
+    $group->id = $groupId;
+    $group->find(TRUE);
+    CRM_Contact_BAO_GroupContactCache::load($group, TRUE);
+
+    // Update contact ACL cache
+    foreach ($contactIds as $contactId) {
+      CRM_ACL_BAO_Cache::updateEntry($contactId);
+    }
 
     CRM_Utils_Hook::post($op, 'GroupContact', $groupId, $contactIds);
 
diff --git a/CRM/Contact/BAO/GroupContactCache.php b/CRM/Contact/BAO/GroupContactCache.php
index a54dcdeb43f..74cf9ae1b41 100644
--- a/CRM/Contact/BAO/GroupContactCache.php
+++ b/CRM/Contact/BAO/GroupContactCache.php
@@ -346,6 +346,7 @@ protected static function flushCaches() {
       // Someone else is kindly doing the refresh for us right now.
       return;
     }
+
     $params = [1 => [self::getCacheInvalidDateTime(), 'String']];
     // @todo this is consistent with previous behaviour but as the first query could take several seconds the second
     // could become inaccurate. It seems to make more sense to fetch them first & delete from an array (which would
@@ -480,30 +481,17 @@ public static function load(&$group, $force = FALSE) {
       return;
     }
 
-    // grab a lock so other processes don't compete and do the same query
-    $lock = Civi::lockManager()->acquire("data.core.group.{$groupID}");
-    if (!$lock->isAcquired()) {
-      // this can cause inconsistent results since we don't know if the other process
-      // will fill up the cache before our calling routine needs it.
-      // however this routine does not return the status either, so basically
-      // its a "lets return and hope for the best"
-      return;
-    }
-
     self::$_alreadyLoaded[$groupID] = 1;
 
-    // we now have the lock, but some other process could have actually done the work
-    // before we got here, so before we do any work, lets ensure that work needs to be
-    // done
+    // FIXME: some other process could have actually done the work before we got here,
+    // Ensure that work needs to be done before continuing
     // we allow hidden groups here since we dont know if the caller wants to evaluate an
     // hidden group
     if (!$force && !self::shouldGroupBeRefreshed($groupID, TRUE)) {
-      $lock->release();
       return;
     }
 
     $sql = NULL;
-    $idName = 'id';
     $customClass = NULL;
     if ($savedSearchID) {
       $ssParams = CRM_Contact_BAO_SavedSearch::getSearchParams($savedSearchID);
@@ -532,7 +520,10 @@ public static function load(&$group, $force = FALSE) {
         if (!strstr($searchSQL, 'WHERE')) {
           $searchSQL .= " WHERE ( 1 ) ";
         }
-        $idName = 'contact_id';
+        $sql = [
+          'select' => substr($searchSQL, 0, strpos($searchSQL, 'FROM')),
+          'from' => substr($searchSQL, strpos($searchSQL, 'FROM')),
+        ];
       }
       else {
         $formValues = CRM_Contact_BAO_SavedSearch::getFormValues($savedSearchID);
@@ -560,56 +551,56 @@ public static function load(&$group, $force = FALSE) {
             FALSE, FALSE,
             FALSE, TRUE
           );
-        $searchSQL = "{$sqlParts['select']} {$sqlParts['from']} {$sqlParts['where']} {$sqlParts['having']} {$sqlParts['group_by']}";
+        $sql = [
+          'select' => $sqlParts['select'],
+          'from' => "{$sqlParts['from']} {$sqlParts['where']} {$sqlParts['having']} {$sqlParts['group_by']}",
+        ];
       }
       $groupID = CRM_Utils_Type::escape($groupID, 'Integer');
-      $sql = $searchSQL . " AND contact_a.id NOT IN (
-                              SELECT contact_id FROM civicrm_group_contact
-                              WHERE civicrm_group_contact.status = 'Removed'
-                              AND   civicrm_group_contact.group_id = $groupID ) ";
+      $sql['from'] .= " AND contact_a.id NOT IN (
+                          SELECT contact_id FROM civicrm_group_contact
+                          WHERE civicrm_group_contact.status = 'Removed'
+                          AND   civicrm_group_contact.group_id = $groupID ) ";
     }
 
-    if ($sql) {
-      $sql = preg_replace("/^\s*SELECT/", "SELECT $groupID as group_id, ", $sql);
+    if (!empty($sql['select'])) {
+      $sql['select'] = preg_replace("/^\s*SELECT/", "SELECT $groupID as group_id, ", $sql['select']);
     }
 
+    $groupContactsTempTable = CRM_Utils_SQL_TempTable::build()->setCategory('gccache')->setMemory();
+    $tempTable = $groupContactsTempTable->getName();
+    $groupContactsTempTable->createWithColumns('contact_id int, group_id int, UNIQUE UI_contact_group (contact_id,group_id)');
+
+    $contactQueries[] = $sql;
     // lets also store the records that are explicitly added to the group
     // this allows us to skip the group contact LEFT JOIN
-    $sqlB = "
-SELECT $groupID as group_id, contact_id as $idName
+    $contactQueries[] = [
+      'select' => "SELECT $groupID as group_id, contact_id as contact_id",
+      'from' => "
 FROM   civicrm_group_contact
 WHERE  civicrm_group_contact.status = 'Added'
-  AND  civicrm_group_contact.group_id = $groupID ";
-
-    self::clearGroupContactCache($groupID);
+  AND  civicrm_group_contact.group_id = $groupID ",
+    ];
 
-    $processed = FALSE;
-    $tempTable = 'civicrm_temp_group_contact_cache' . rand(0, 2000);
-    foreach ([$sql, $sqlB] as $selectSql) {
-      if (!$selectSql) {
+    foreach ($contactQueries as $contactQuery) {
+      if (empty($contactQuery['select']) || empty($contactQuery['from'])) {
         continue;
       }
-      $insertSql = "CREATE TEMPORARY TABLE $tempTable ($selectSql);";
-      $processed = TRUE;
-      CRM_Core_DAO::executeQuery($insertSql);
-      CRM_Core_DAO::executeQuery(
-        "INSERT IGNORE INTO civicrm_group_contact_cache (contact_id, group_id)
-        SELECT DISTINCT $idName, group_id FROM $tempTable
-      ");
-      CRM_Core_DAO::executeQuery(" DROP TEMPORARY TABLE $tempTable");
+      if (CRM_Core_DAO::singleValueQuery("SELECT COUNT(*) {$contactQuery['from']}") > 0) {
+        CRM_Core_DAO::executeQuery("INSERT IGNORE INTO $tempTable (contact_id, group_id) {$contactQuery['select']} {$contactQuery['from']}");
+      }
     }
 
-    self::updateCacheTime([$groupID], $processed);
-
     if ($group->children) {
 
-      //Store a list of contacts who are removed from the parent group
-      $sql = "
+      // Store a list of contacts who are removed from the parent group
+      $sqlContactsRemovedFromGroup = "
 SELECT contact_id
 FROM civicrm_group_contact
 WHERE  civicrm_group_contact.status = 'Removed'
 AND  civicrm_group_contact.group_id = $groupID ";
-      $dao = CRM_Core_DAO::executeQuery($sql);
+
+      $dao = CRM_Core_DAO::executeQuery($sqlContactsRemovedFromGroup);
       $removed_contacts = [];
       while ($dao->fetch()) {
         $removed_contacts[] = $dao->contact_id;
@@ -618,19 +609,52 @@ public static function load(&$group, $force = FALSE) {
       $childrenIDs = explode(',', $group->children);
       foreach ($childrenIDs as $childID) {
         $contactIDs = CRM_Contact_BAO_Group::getMember($childID, FALSE);
-        //Unset each contact that is removed from the parent group
+        // Unset each contact that is removed from the parent group
         foreach ($removed_contacts as $removed_contact) {
           unset($contactIDs[$removed_contact]);
         }
+        if (empty($contactIDs)) {
+          // This child group has no contact IDs so we don't need to add them to
+          continue;
+        }
         $values = [];
         foreach ($contactIDs as $contactID => $dontCare) {
           $values[] = "({$groupID},{$contactID})";
         }
-
-        self::store([$groupID], $values);
+        $str = implode(',', $values);
+        CRM_Core_DAO::executeQuery("INSERT IGNORE INTO $tempTable (group_id, contact_id) VALUES $str");
       }
     }
 
+    // grab a lock so other processes don't compete and do the same query
+    $lock = Civi::lockManager()->acquire("data.core.group.{$groupID}");
+    if (!$lock->isAcquired()) {
+      // this can cause inconsistent results since we don't know if the other process
+      // will fill up the cache before our calling routine needs it.
+      // however this routine does not return the status either, so basically
+      // its a "lets return and hope for the best"
+      return;
+    }
+
+    // Don't call clearGroupContactCache as we don't want to clear the cache dates
+    // The will get updated by updateCacheTime() below and not clearing the dates reduces
+    // the chance that loadAll() will try and rebuild at the same time.
+    $clearCacheQuery = "
+    DELETE  g
+      FROM  civicrm_group_contact_cache g
+      WHERE  g.group_id = %1 ";
+    $params = [
+      1 => [$groupID, 'Integer'],
+    ];
+    CRM_Core_DAO::executeQuery($clearCacheQuery, $params);
+
+    CRM_Core_DAO::executeQuery(
+      "INSERT IGNORE INTO civicrm_group_contact_cache (contact_id, group_id)
+        SELECT DISTINCT contact_id, group_id FROM $tempTable
+      ");
+    $groupContactsTempTable->drop();
+    self::updateCacheTime([$groupID], TRUE);
+
     $lock->release();
   }
 
